c7n/resources/ami.py

```diff
@@ -313,19 +336,33 @@ def _pull_asg_images(self):
             image_ids.update([
                 lcfg['ImageId'] for lcfg in lcfg_mgr.resources()
                 if lcfg['LaunchConfigurationName'] in lcfgs])
+        elif not lcfgs and include_lcfg:
+            image_ids.update([lcfg['ImageId'] for lcfg in lcfg_mgr.resources()])
 
         tmpl_mgr = self.manager.get_resource_manager('launch-template-version')
-        for tversion in tmpl_mgr.get_resources(
-                list(tmpl_mgr.get_asg_templates(asgs).keys())):
-            image_ids.add(tversion['LaunchTemplateData'].get('ImageId'))
+        if asgs or include_lcfg:
+            for tversion in tmpl_mgr.get_resources(
+                    list(tmpl_mgr.get_asg_templates(asgs).keys())):
```

Comment:
it seems like some of this is a reimplementation of the logic already in asg LaunchInfo ie. `LauchInfo(manager).initialize(asgs).get_image_ids()`

Reply:
there's actually a delta in behavior between LaunchInfo and the implementation here, LaunchInfo only returns launch configs/templates that are asg attached, see `LaunchInfo.get_launch_templates` and `LaunchInfo.get_launch_configs`. This returns all launch configs and templates in the case that the user specifies that they want to query all configs/templates regardless of if it's asg attached.